components/director/internal/systemfetcher/service.go

```diff
@@ -227,21 +239,75 @@ func (s *SystemFetcher) processSystemsForTenant(ctx context.Context, tenantMappi
 	return nil
 }
 
-func (s *SystemFetcher) convertSystemToAppRegisterInput(sc System) (model.ApplicationRegisterInput, string) {
-	initStatusCond := model.ApplicationStatusConditionInitial
-	baseURL := sc.BaseURL
+func (s *SystemFetcher) convertSystemToAppRegisterInput(ctx context.Context, sc System) (*model.ApplicationRegisterInputWithTemplate, error) {
+	if len(sc.TemplateID) > 0 {
+		return s.appRegisterInputFromTemplate(ctx, sc)
+	}
 
 	appRegisterInput := model.ApplicationRegisterInput{
-		Name:            sc.DisplayName,
-		Description:     &sc.ProductDescription,
-		BaseURL:         &baseURL,
-		ProviderName:    &sc.InfrastructureProvider,
-		StatusCondition: &initStatusCond,
-		SystemNumber:    &sc.SystemNumber,
-		Labels: map[string]interface{}{
-			"managed": "true",
-		},
-	}
-
-	return appRegisterInput, sc.TemplateID
+		Name: sc.DisplayName,
+	}
+
+	return enrichAppRegisterInput(appRegisterInput, sc), nil
+}
+
+func (s *SystemFetcher) appRegisterInputFromTemplate(ctx context.Context, sc System) (*model.ApplicationRegisterInputWithTemplate, error) {
+	appTemplate, err := s.appTemplateService.Get(ctx, sc.TemplateID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting application template with ID %s", sc.TemplateID)
+	}
+
+	inputValues := getTemplateInputs(appTemplate, sc)
+	appRegisterInputJSON, err := s.appTemplateService.PrepareApplicationCreateInputJSON(appTemplate, inputValues)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while preparing ApplicationRegisterInput JSON from Application Template with name %s", appTemplate.Name)
+	}
+
+	appRegisterInput, err := s.appConverter.CreateInputJSONToModel(ctx, appRegisterInputJSON)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while preparing ApplicationRegisterInput model from Application Template with name %s", appTemplate.Name)
+	}
+
+	return enrichAppRegisterInput(appRegisterInput, sc), nil
+}
+
+func enrichAppRegisterInput(input model.ApplicationRegisterInput, sc System) *model.ApplicationRegisterInputWithTemplate {
+	initStatusCond := model.ApplicationStatusConditionInitial
+
+	input.Name = sc.DisplayName
+	input.Description = &sc.ProductDescription
+	input.BaseURL = &sc.BaseURL
+	input.ProviderName = &sc.InfrastructureProvider
+	input.StatusCondition = &initStatusCond
+	input.SystemNumber = &sc.SystemNumber
+
+	if len(input.Labels) == 0 {
+		input.Labels = make(map[string]interface{}, 1)
+	}
+	input.Labels["managed"] = "true"
+
+	return &model.ApplicationRegisterInputWithTemplate{
+		ApplicationRegisterInput: input,
+		TemplateID:               sc.TemplateID,
+	}
+}
+
+func getTemplateInputs(t *model.ApplicationTemplate, s System) model.ApplicationFromTemplateInputValues {
```

Comment:
This function populates the placeholders just for the sake of the service layer to work and add all the static (directly hardcoded things) like type labels from the template.

As such I think we have two options:

1. Either mark it as a fully dummy/workaround function (and document it as such with a godoc) and then assume that everything will be populated by the enrich function directly.
2. Another option is to let it at least try to resolve the fields it can. The best way I can think of is to parse the json tags of the System struct and then search for them in the placeholder. If a similar placeholder is found we then can use the value to populate it. For all placeholders that do not match any json tag we can use the default value (ex. DisplayName)

If we choose option two we should choose what we do if a template matches. Should we identify this in the enrich or the properties we have there are always overridden as right now.

My impression is regarding the `url` "guess". I think that we should either abstain on any trials for real populating the placeholders or do whatever we can in order to make it usable in the future (if we use json tags we know what a system will contain and we can extend the templates to use it - not sure if it actually matters if we override it anyways in the `enrichAppRegisterInput`)

Reply:
Or we can improve the placeholders, and add a new field which will hold the mapping to a field of the systems fetched from an external source?

Reply:
I agree now that the templates make sense to be used in this case. 

What do you think about going all-in for the templates and change the templates to match the pattern of the systems from an external source?
This means that the `enrichAppRegisterInput` will no longer be needed and the template will take care of that.

However, with this approach, we should add one more field for each placeholder: `optional bool`
Where all external system-specific properties will be optional and therefore ignored in the standard case.

This will make the templates a bit more usable because the Cockpit can be extended to generate a form based on the placeholders and if they are optional or required. Then when a person registers a system from a template can add more metadata to the id than the name (like systemNumber etc..) if he knows about this stuff.

The system fetcher will also be simplified because it will work with the template only and will not assume a priority of the information sources.